pkg/platform/packet/packet.go

```diff
@@ -153,16 +154,15 @@ func (c *config) Apply(ex *terraform.Executor) error {
 
 	c.AssetDir = assetDir
 
-	dnsProvider, err := dns.ParseDNS(&c.DNS)
-	if err != nil {
+	if err := dns.Validate(c.DNSProvider); err != nil {
 		return errors.Wrap(err, "parsing DNS configuration failed")
 	}
 
 	if err := c.Initialize(ex); err != nil {
 		return err
 	}
 
-	return c.terraformSmartApply(ex, dnsProvider)
+	return c.terraformSmartApply(ex, c.DNSProvider, c.DNSZone)
```

Comment:
Do you see a need for DNS-related parameters other than provider and zone? IMO we shouldn't add an abstraction just because at some undefined point in the future we _may_ want to add some parameter somewhere. You can use this "future changes" argument nearly always, however until you have a concrete problem (e.g. "we constantly have to modify the function signature of `terraformSmartApply`"), more likely than not you will produce an abstraction that isn't flexible or doesn't serve the purpose the way you initially thought.

Reply:
In my opinion,
```go
return c.terraformSmartApply(ex, dnsProvider)
```
is better than
```go
return c.terraformSmartApply(ex, c.DNSProvider, c.DNSZone)
```
because:
- it is less error prone, as order of parameters does not matter anymore
- less function parameters is usually better
- it is just less code
- in this particular case, it is less diff, which makes it easier to review
- it allows adding more DNS configuration fields transparent (no need to change the function signature)
- if the struct is used for DNS configuration externally and this function stays like this, it saves expanding parameters (so `Config{}` -> `(c.DNSProvider, c.DNSZone)`)
- for static calls, you don't have `terraformSmartApply(terraformExecutor, <some string>, <some string>)`, but `terraformSmartApply(terraformExecutor, dnsConfiguration)`, so the code has more meaning
- if we add another `string` parameter, or in the middle, or after Terraform executor, it won't make a difference for the compiler, so it may potentially introduce bugs

In my opinion, not even trying to predict of what is going to happen with this code is a poor choice, as I'm 100% sure it will require changes at some point in the future. The abstraction at this level is cheap to build, small and can be easily changed when there is a demand. It also potentially saves time for the future engineer which will work on it, as right now this is a **single task** to implement it this way. In the future if someone wants to change it, they will have to first make it changable and then add their desired change, which is **two tasks**.

Reply:
>In my opinion, not even trying to predict of what is going to happen with this code is a poor choice, as I'm 100% sure it will require changes at some point in the future. The abstraction at this level is cheap to build, small and can be easily changed when there is a demand. It also potentially saves time for the future engineer which will work on it, as right now this is a single task to implement it this way. In the future if someone wants to change it, they will have to first make it changable and then add their desired change, which is two tasks.

The reality is you can't predict the future. I believe that building an abstraction before it is necessary tends to do more harm than good. The problem is not the cost of building an abstraction, but rather the cost of _dismantling_ the wrong abstraction, which can be a real pain especially as implementations grow in complexity. True, I cannot know this is the wrong abstraction at this point, but I also cannot know it's the right abstraction, and that's why I'd rather wait with creating it until there is a tangible need for it because then we'll be a lot smarter about this.

Anyway, although I'm not convinced by the reasoning above, I've made the change as you suggested for the sake of keeping things productive. Please keep in mind this turns `dns` into a block at the config level, too.